### PR TITLE
Mapped field types searchable with doc values

### PR DESCRIPTION
**Pull request by tmgordeeva**
When using TSDB, time series metrics aren't indexed but do have doc values. Field caps should report those fields as searchable.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java [hunk 1]
```diff
@@ -148,7 +148,7 @@ public Object valueForDisplay(Object value) {
      * Returns true if the field is searchable.
      */
     public boolean isSearchable() {
-        return isIndexed;
+        return isIndexed || hasDocValues();
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:151]: I think this needs to be made in `ScaledFloatFieldType` by overwriting `isSearchable()` method.
Or maybe let `ScaledFloatFieldType` extend from `NumberFieldType`?

**Comment by tmgordeeva** [comment]
@salvatore-campagna Could you review this while Martinj is out? This is for https://github.com/elastic/elasticsearch/issues/97394

**Comment by salvatore-campagna** [comment]
Why is this done only for `scaled float`? I would have expected this to be done for all metrics (gauge and counters) no matter the actual specific type. If I recall correctly we do not index metrics because the expectation is that we aggregate metrics and we do not search on them. Of course, anyway, we would like to support searching through doc values. For instance, what happens if the metric is declared float or long or whatever other type? Am I missing something?

**Comment by tmgordeeva** [comment]
This is the same logic as in `NumberFieldType` - originally I wanted to inherit off of that class, but it has enough different logic in `termsQuery` and other shared methods it seemed not worth doing a refactoring that could just make both classes more confusing.

Is it safe to apply this logic to every field type? We don't do the `hasDocValues` check for all field types, but I'm not entirely sure why that is - I'll check with search for more information.

**Comment by tmgordeeva** [comment]
This is the same logic as in `NumberFieldType` - which as Martinj suggested we wanted to inherit from, but there's enough conflicting logic I think trying to force inheritance just makes both classes more confusing. Do the other gauges/counters use `NumberFieldType`?

I'm also going to check with search exactly why we don't always check `hasDocValues()`.

**Comment by martijnvg** [comment]
I don't know whether why `ScaledFloatFieldType` doesn't extend from `NumberFieldType`, but if doing this cause other issues then let's not do this for now, and just make the change in the `ScaledFloatFieldType#isSearchable()` method.

> I'm also going to check with search exactly why we don't always check hasDocValues().

I think this because not all field types are searchable when only doc values are enabled. So this change can't be made to for example `SimpleMappedFieldType`.

For all the base number fields that can be a gauge or counter the right thing happens now because these fields use the `NumberFieldType` class. Also for fields of type `GeoPointFieldType`, it is searchable if either it is indexed or doc values are enabled.   However I did notice that for fields of type `UnsignedLongFieldType` and  `AggregateDoubleMetricFieldType`, are not searchable if only doc values is enabled and I think that is wrong (similar to scaled_float which is addressed in this pr). I think similar fixes need to be made to `UnsignedLongFieldType` and  `AggregateDoubleMetricFieldType`. This can be done in this PR or followup PRs.

File: modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java [hunk 2]
```diff
@@ -156,6 +156,22 @@ public void testNoDocValues() throws Exception {
...
+    public void testDocValuesUnsearchable() throws Exception {
+        MapperService mapperService = createMapperService(
+            fieldMapping(
+                b -> b.field("type", "scaled_float").field("index", false).field("doc_values", false).field("scaling_factor", 10.0)
+            )
+        );
+        MappedFieldType fieldType = mapperService.fieldType("field");
+        assertFalse(fieldType.isSearchable());
+    }
+
```
Comment (martijnvg) [inline modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapperTests.java:174]: Maybe also add unit tests for the following cases: 

- index=false, doc_values=true and assert that field is searchable.
- index=true, doc_values=false and assert that field is searchable.
- defaults and time_series_metric=gauge and assert that field is searchable.

**Review by martijnvg — APPROVED** [review]
Left a minor testing comment, otherwise LGTM. Like I mentioned in the other test, I think we have the same issue with unsigned long and aggregate double metric field type. But I think we can address that in different PRs. This we can then backport to the 8.9 branch, so it can be included in the next bug fix release.

**Pull request merged**